synapse/storage/databases/main/deviceinbox.py

```diff
@@ -218,6 +218,9 @@ async def get_messages_for_device(
                 * The last-processed stream ID. Subsequent calls of this function with the
                   same device should pass this value as 'from_stream_id'.
         """
+        if limit == 0:
+            return [], from_stream_id
```

Comment:
```suggestion
            # We return the from token so that if a sync later on asks for
            # non-zero number of to-device messages we won't have dropped any.
            return [], from_stream_id
```